### PR TITLE
ciri: network refactoring

### PR DESCRIPTION
**Pull request by thibault-martinez**
- Drops UDP support
- Removes protocol field from endpoint struct
- Removes boost-based services
- Uses https://github.com/libuv/libuv for high efficiency network
- Adds new configuration variables (`--auto-tethering-enabled`, `--max-neighbors`, `--neighboring-address`, `--neighboring-port` and `--reconnect-attempt-interval`)
- Adds a `protocol` folder and move `iota_packet` and `transaction_request` request inside
- Adds dropped packets metric to neighbors
- Adds a `_stage` suffix to pipeline components
- Fixes request hash size to 49 for both mainnet and testnet
- Adds a libuv-based TCP server
- Adds a TLV protocol with types `header`, `handshake` and `gossip`
- Renames `iota_packet_t` to `protocol_gossip_t`
- Adds a router that handles neighbors
- Moves neighbors to network folder
- Txs broadcasted by the API are now also part of the processing pipeline
- Removes useless cond timers
- Adds state and protocol version to neighbors
- Renames endpoint host to domain
- Adds handshake packet
- Adds packet types
- Adds TCP DNS resolver
- Receives and initiates connections
- Adds periodic reconnect attempts

### CODE REVIEW

File: ciri/node/network/neighbor.c [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+  return RC_OK;
+}
+
+retcode_t neighbor_send_packet(node_t *const node, neighbor_t *const neighbor, protocol_gossip_t const *const packet) {
+  if (node == NULL || neighbor == NULL || packet == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  neighbor->nbr_sent_txs++;
```
Comment (tsvisabo) [inline ciri/node/network/neighbor.c]: if this function only increments a counter should it even be a function?
also, why's the prior check?

File: ciri/node/network/neighbor.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  uint64_t nbr_random_tx_reqs;
+  uint64_t nbr_sent_txs;
+  uint64_t nbr_new_txs;
+  uint64_t nbr_dropped_send;
+} neighbor_t;
+
+/**
+ * Initializes a neighbor with an URI
+ *
+ * @param neighbor The neighbor
```
Comment (tsvisabo) [inline ciri/node/network/neighbor.h]: [in, out] ?

File: ciri/node/network/router.c [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+  }
+
+  return endpoint_cmp(&((neighbor_t *)lhs)->endpoint, &((neighbor_t *)rhs)->endpoint);
+}
+
+static retcode_t router_neighbors_init(router_t *const router) {
+  neighbor_t neighbor;
+  char *neighbor_uri, *cpy, *ptr;
+
+  if (router->node->conf.neighbors != NULL) {
```
Comment (tsvisabo) [inline ciri/node/network/router.c:38]: I think we should log a warning if there are no neighbors?

File: common/errors.h [hunk 4]
```diff
@@ -327,6 +328,15 @@ enum retcode_t {
...
   RC_LOCK_DESTROY = 0x07 | RC_MODULE_HANDLE | RC_SEVERITY_FATAL,
+
+  // Network Module
+  RC_DOMAIN_RESOLUTION = 0x01 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
+  RC_EVENT_LOOP = 0x02 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
+  RC_TCP_SERVER_INIT = 0x03 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
+  RC_INVALID_PACKET = 0x04 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
+  RC_INVALID_PACKET_TYPE = 0x05 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
+  RC_FAILED_HANDSHAKE = 0x06 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
+  RC_WRITE_FAILED = 0x07 | RC_MODULE_NETWORK | RC_SEVERITY_FATAL,
```
Comment (tsvisabo) [inline common/errors.h]: sounds too generic

File: ciri/node/protocol/type.h [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+#include "ciri/node/protocol/gossip.h"
+#include "ciri/node/protocol/handshake.h"
+#include "ciri/node/protocol/header.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define PACKET_TYPE_NBR 3
```
Comment (tsvisabo) [inline ciri/node/protocol/type.h]: PACKET_TYPES_NUM? : []

File: ciri/node/protocol/handshake.h [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+#define HANDSHAKE_SUPPORTED_VERSIONS_MAX_BYTES_LENGTH 32
+#define HANDSHAKE_MIN_BYTES_LENGTH                                                                                     \
+  ((HANDSHAKE_PORT_BYTES_LENGTH) + (HANDSHAKE_TIMESTAMP_BYTES_LENGTH) + (HANDSHAKE_COORDINATOR_ADDRESS_BYTES_LENGTH) + \
+   (HANDSHAKE_MWM_BYTES_LENGTH) + (HANDSHAKE_SUPPORTED_VERSIONS_MIN_BYTES_LENGTH))
+#define HANDSHAKE_MAX_BYTES_LENGTH                                                                                     \
+  ((HANDSHAKE_PORT_BYTES_LENGTH) + (HANDSHAKE_TIMESTAMP_BYTES_LENGTH) + (HANDSHAKE_COORDINATOR_ADDRESS_BYTES_LENGTH) + \
+   (HANDSHAKE_MWM_BYTES_LENGTH) + (HANDSHAKE_SUPPORTED_VERSIONS_MAX_BYTES_LENGTH))
+
+#define HANDSHAKE_PROTOCOL_VERSION = 1
+static uint8_t handshake_supported_protocol_versions[] = {0b00000001};
```
Comment (tsvisabo) [inline ciri/node/protocol/handshake.h:33]: is that correct?

File: ciri/node/protocol/handshake.h [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  uint64_t timestamp;
+  uint8_t coordinator_address[HANDSHAKE_COORDINATOR_ADDRESS_BYTES_LENGTH];
+  uint8_t mwm;
+  uint8_t supported_versions[HANDSHAKE_SUPPORTED_VERSIONS_MAX_BYTES_LENGTH];
+} protocol_handshake_t;
+
+/**
+ * Initializes an handshake packet
+ *
+ * @param handshake The handshake packet
```
Comment (tsvisabo) [inline ciri/node/protocol/handshake.h]: [in, out] ?

File: ciri/node/protocol/handshake.c [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  *handshake_size = HANDSHAKE_MIN_BYTES_LENGTH + sizeof(handshake_supported_protocol_versions) - 1;
+}
+
+int handshake_supported_version(uint8_t const* const own_supported_versions, size_t const own_supported_versions_length,
+                                uint8_t const* const supported_versions, size_t const supported_versions_length) {
+  int highest_supported_version = 0;
+
+  for (size_t i = 0; i < own_supported_versions_length; i++) {
+    // max check up to advertised versions by the neighbor
+    if (i > supported_versions_length - 1) {
```
Comment (tsvisabo) [inline ciri/node/protocol/handshake.c]: can you remove that `if` by taking the MIN(own_supported_versions_length,supported_versions_length - 1) and plug it in in the `for` condition?

File: ciri/node/protocol/gossip.h [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+
+#include "utlist.h"
+
+#include "ciri/node/network/endpoint.h"
+#include "common/errors.h"
+#include "common/trinary/bytes.h"
+#include "common/trinary/flex_trit.h"
+
+// The amount of bytes making up the non signature message fragment part of a transaction gossip payload
+#define GOSSIP_NON_SIG_BYTES_LENGTH 292
```
Comment (tsvisabo) [inline ciri/node/protocol/gossip.h:19]: should we call that `GOSSIP_ESSENCE_BYTES_LENGTH` instead?

File: ciri/node/protocol/gossip.h [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+ * A queue of gossip packets used to dispatch packets in different threads.
+ * Not concurrent by default.
+ */
+typedef struct protocol_gossip_queue_entry_s {
+  protocol_gossip_t packet;
+  struct protocol_gossip_queue_entry_s* next;
+  struct protocol_gossip_queue_entry_s* prev;
+} protocol_gossip_queue_entry_t;
+
+typedef protocol_gossip_queue_entry_t* protocol_gossip_queue_t;
```
Comment (tsvisabo) [inline ciri/node/protocol/gossip.h:56]: Can we template a container instead?

File: ciri/node/protocol/gossip.h [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+/**
+ * Sets the transaction of a gossip packet
+ *
+ * @param packet The gossip packet
```
Comment (tsvisabo) [inline ciri/node/protocol/gossip.h]: [in, out] ?

File: ciri/node/network/router.h [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+typedef struct router_s {
+  node_t *node;
+  UT_array *neighbors;
+  rw_lock_handle_t neighbors_lock;
+} router_t;
+
+/**
+ * Initializes a router
+ *
+ * @param router The router
```
Comment (tsvisabo) [inline ciri/node/network/router.h]: [in , out]

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: ciri/node/network/neighbor.c (hunk 1 above)
Comment (tsvisabo) [reply ciri/node/network/neighbor.c]: ah, there should have been a `TODO` there, cool!

File: ciri/node/network/neighbor.c [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+  if (node == NULL || neighbor == NULL || neighbor->endpoint.stream == NULL || packet == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  for (int i = GOSSIP_SIG_MAX_BYTES_LENGTH - 1; i >= 0 && packet->content[i] == 0; i--) {
+    content_length--;
+  }
+
+  memcpy(content, packet->content, content_length);
```
Comment (tsvisabo) [inline ciri/node/network/neighbor.c]: Could we instead reuse `packet` content and truncate `essence` + `requested_hash` to the end of `sig`,
i.e `memcpy(&packet->content[content_length], packet->content + GOSSIP_SIG_MAX_BYTES_LENGTH, GOSSIP_NON_SIG_BYTES_LENGTH + GOSSIP_REQUESTED_TX_HASH_BYTES_LENGTH)`?
(not sure that's a safe way though, but you get the point)

File: ciri/node/network/router.c [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+  // We haven't received a full header yet
+  if (neighbor->buffer_size <= HEADER_BYTES_LENGTH) {
+    return RC_OK;
+  }
+
+  header = (protocol_header_t const *)neighbor->buffer;
+  header_length = ntohs(header->length);
+
+  // We haven't received the full packet yet
+  if (neighbor->buffer_size - HEADER_BYTES_LENGTH < header_length) {
```
Comment (tsvisabo) [inline ciri/node/network/router.c:614]: should it not be `if (neighbor->buffer_size < HEADER_BYTES_LENGTH)`?
i don't understand why you compare the size of the part of buffer which is not header to header..

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: ciri/node/network/neighbor.c (hunk 1 above)
Comment (thibault-martinez) [reply ciri/node/network/neighbor.c]: Yep, sorry about that.

File: ciri/node/network/router.c (hunk 3 above)
Comment (thibault-martinez) [reply ciri/node/network/router.c:38]: Good idea

File: ciri/node/protocol/type.h (hunk 5 above)
Comment (thibault-martinez) [reply ciri/node/protocol/type.h]: 😬

File: ciri/node/protocol/handshake.h (hunk 6 above)
Comment (thibault-martinez) [reply ciri/node/protocol/handshake.h:33]: It is, but I don't like it either. I'll find a way to change.

File: ciri/node/network/router.c (hunk 14 above)
Comment (thibault-martinez) [reply ciri/node/network/router.c:614]: `header_length` contains the advertised length in the header. It indicates how much data is coming after. So in the buffer we need to have at least this amount.

File: ciri/node/protocol/gossip.h (hunk 10 above)
Comment (thibault-martinez) [reply ciri/node/protocol/gossip.h:56]: https://github.com/iotaledger/entangled/issues/1340

File: ciri/node/protocol/handshake.c (hunk 8 above)
Comment (thibault-martinez) [reply ciri/node/protocol/handshake.c]: Sure, but `MIN(own_supported_versions_length,supported_versions_length)` though ?

File: ciri/node/protocol/gossip.h (hunk 9 above)
Comment (thibault-martinez) [reply ciri/node/protocol/gossip.h:19]: I agree we maybe find another name but technically it's not the essence, it's all the fields except the signature.

File: ciri/node/network/neighbor.c (hunk 13 above)
Comment (thibault-martinez) [reply ciri/node/network/neighbor.c]: That would mean that sending a packet modifies it ?

File: common/errors.h (hunk 4 above)
Comment (thibault-martinez) [reply common/errors.h]: How would you call it ?

Comment (tsvisabo) [reply common/errors.h]: RC_WRITE_TCP_FAILED ?

File: ciri/node/protocol/gossip.h (hunk 9 above)
Comment (tsvisabo) [reply ciri/node/protocol/gossip.h:19]: If so, current name is the best

File: ciri/node/network/neighbor.c (hunk 13 above)
Comment (tsvisabo) [reply ciri/node/network/neighbor.c]: yes, is it going to be used anywhere?
also, if the sig is full, no need to do memcpy at all

Comment (tsvisabo) [reply ciri/node/network/neighbor.c]: idk, is that an important performance gain in your opinion?

File: common/errors.h (hunk 4 above)
Comment (thibault-martinez) [reply common/errors.h]: Deal

File: ciri/node/network/neighbor.c [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+  int err = 0;
+
+  memset(neighbor, 0, sizeof(neighbor_t));
+  neighbor->state = NEIGHBOR_DISCONNECTED;
+  neighbor->endpoint.stream = NULL;
+  lock_handle_init(&neighbor->buffer_lock);
+  if ((neighbor->writer = (uv_async_t *)malloc(sizeof(uv_async_t))) == NULL) {
+    return RC_OOM;
+  }
+  if ((err = uv_async_init(uv_default_loop(), neighbor->writer, tcp_server_on_async_write)) != 0) {
```
Comment (tsvisabo) [inline ciri/node/network/neighbor.c]: ?
need to free `neighbor->writer`

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: ciri/node/network/neighbor.c (hunk 13 above)
Comment (thibault-martinez) [reply ciri/node/network/neighbor.c]: Not relevant anymore

File: ciri/node/network/neighbor.c (hunk 15 above)
Comment (thibault-martinez) [reply ciri/node/network/neighbor.c]: It is freed by a callback

File: ciri/node/network/neighbor.c [hunk 16]
```diff
@@ -14,17 +14,30 @@
...
 
+retcode_t neighbor_init(neighbor_t *const neighbor) {
+  if (neighbor == NULL) {
+    return RC_NULL_PARAM;
+  }
+
+  memset(neighbor, 0, sizeof(neighbor_t));
+  neighbor->writer = NULL;
+  neighbor->endpoint.stream = NULL;
+  neighbor->state = NEIGHBOR_DISCONNECTED;
```
Comment (tsvisabo) [inline ciri/node/network/neighbor.c:25]: what about `lock_handle_init(&neighbor->buffer_lock);`?

File: ciri/node/network/router.c [hunk 17]
```diff
@@ -106,9 +107,23 @@ retcode_t router_neighbor_add(router_t *const router, neighbor_t *const neighbor
     ret = RC_NEIGHBOR_ALREADY_PAIRED;
   } else {
     utarray_push_back(router->neighbors, neighbor);
+    elt = (neighbor_t *)utarray_back(router->neighbors);
+    lock_handle_init(&elt->buffer_lock);
```
Comment (tsvisabo) [inline ciri/node/network/router.c]: should this not be part of neighbor_init?

File: ciri/node/network/tcp_server.c [hunk 18]
```diff
@@ -415,20 +415,26 @@ retcode_t tcp_server_write(uv_stream_t *const stream, packet_type_t const type,
...
+  rw_lock_handle_wrlock(&neighbor->write_queue_lock);
+  entry = neighbor_write_queue_pop(neighbor);
+  rw_lock_handle_unlock(&neighbor->write_queue_lock);
+
+  if (entry != NULL) {
+    req->data = entry->buf.base;
+    if ((err = uv_write(req, neighbor->endpoint.stream, &entry->buf, 1, tcp_server_on_write)) != 0) {
+      log_warning(logger_id, "Writing failed: %s\n", uv_err_name(err));
+    }
+    neighbor->nbr_sent_txs++;
```
Comment (tsvisabo) [inline ciri/node/network/tcp_server.c]: should that be incremented if `err != RC_OK` ?

**Review by tsvisabo — CHANGES_REQUESTED** [review]

File: ciri/node/network/tcp_server.c (hunk 18 above)
Comment (thibault-martinez) [reply ciri/node/network/tcp_server.c]: Definitely

**Review by tsvisabo — DISMISSED** [review]

**Review by tsvisabo — APPROVED** [review]

**Pull request merged**